Plugins/GitUIPluginInterfaces/ObjectId.cs

```diff
@@ -0,0 +1,178 @@
+    /// Models a SHA1 hash.
+    /// </summary>
+    /// <remarks>
+    /// Instances are immutable and are guaranteed to contain valid, 160-bit (20-byte) SHA1 hashes.
+    /// </remarks>
+    public sealed class ObjectId : IEquatable<ObjectId>
+    {
+        private static readonly Random _random = new Random();
+
+        private const int Sha1ByteCount = 20;
+        private const int Sha1CharCount = 40;
+
+        public static ObjectId UnstagedId { get; } = Parse("0000000000000000000000000000000000000000");
+        public static ObjectId IndexId    { get; } = Parse("1111111111111111111111111111111111111111");
+
+        public static ObjectId Random()
+        {
+            var bytes = new byte[Sha1ByteCount];
+            _random.NextBytes(bytes);
+            return new ObjectId(bytes);
+        }
+
+        #region Parsing
+
+        public static bool TryParse([CanBeNull] string hex, out ObjectId objectId)
+        {
+            if (hex == null || hex.Length != Sha1CharCount)
+            {
+                objectId = default;
+                return false;
+            }
+
+            return TryParse(hex, 0, out objectId);
+        }
+
+        public static bool TryParse([CanBeNull] string hex, int offset, out ObjectId objectId)
+        {
+            if (hex == null || hex.Length - offset < Sha1CharCount)
+            {
+                objectId = default;
+                return false;
+            }
+
+            var bytes = new byte[Sha1ByteCount];
+
+            var charIndex = 0;
+            for (var i = 0; i < Sha1ByteCount; ++i)
+            {
+                var nibble1 = HexCharToInt(hex[charIndex++]);
+                var nibble2 = HexCharToInt(hex[charIndex++]);
+
+                if (nibble1 < 0 || nibble1 > 15 || nibble2 < 0 || nibble2 > 15)
+                {
+                    objectId = default;
+                    return false;
+                }
+
+                bytes[i] = (byte)((nibble1 << 4) + nibble2);
+            }
+
+            objectId = new ObjectId(bytes);
+            return true;
+
+            int HexCharToInt(char c)
+            {
+                var sub = c <= '9'
+                    ? '0'
+                    : c < 'a'
+                        ? 'A' - 10
+                        : 'a' - 10;
+
+                return c - sub;
+            }
+        }
+
+        public static ObjectId Parse(string hex)
+        {
+            if (hex.Length != Sha1CharCount || !TryParse(hex, 0, out var id))
+                throw new FormatException($"Unable to parse object ID \"{hex}\".");
+            return id;
+        }
+
+        public static ObjectId Parse(string hex, int offset)
+        {
+            if (!TryParse(hex, offset, out var id))
+                throw new FormatException($"Unable to parse object ID \"{hex}\".");
+            return id;
+        }
+
+        #endregion
+
+        public static bool IsValid(string hex)
+        {
+            if (hex.Length != Sha1CharCount)
```

Comment:
Git can take variable length SHAs.
Why are forcing 40 chars here?

Same applies to `TryParse`